lib/auth/keystore/hsm.go

```diff
@@ -0,0 +1,303 @@
+		ctx:      ctx,
+		hostUUID: config.HostUUID,
+		log:      logrus.WithFields(logrus.Fields{trace.Component: "HSMKeyStore"}),
+	}, nil
+}
+
+// GenerateRSA creates a new RSA private key and returns its identifier and a
+// crypto.Signer. The returned identifier can be passed to GetSigner later to
+// get the same crypto.Signer.
+func (c *hsmKeyStore) GenerateRSA() ([]byte, crypto.Signer, error) {
+	var id uuid.UUID
+	var err error
+
+	// Some HSMs (like YubiHSM2) will silently truncate the passed ID to as few
+	// as 2 bytes. There's not a great way to detect this and I don't want to
+	// limit the ID to 2 bytes on all systems, so for now we will generate a
+	// few random IDs and hope to avoid a collision. Ideally Teleport should be
+	// the only thing creating keys for this token and there should only be 10
+	// keys per HSM at a given time:
+	// 2(rotation phases) * (4(SSH and TLS for User and Host CA) + 1(JWT CA))
+	maxIterations := 16
+	iterations := 0
+	for ; iterations < maxIterations; iterations++ {
+		id, err = uuid.NewRandom()
+		if err != nil {
+			return nil, nil, trace.Wrap(err)
+		}
+		existingSigner, err := c.ctx.FindKeyPair(id[:], label)
+		// some HSMs (like YubiHSM2) will return signer == nil && err == nil if
+		// no match is found
+		if err != nil || existingSigner == nil {
+			// failed to find an existing keypair, so this ID is unique
+			break
+		} else {
+			c.log.Warn("Found CKA_ID collision while creating keypair, retrying with new ID")
+		}
+	}
+	if iterations == maxIterations {
+		return nil, nil, trace.AlreadyExists("failed to find unused CKA_ID for HSM")
+	}
+
+	signer, err := c.ctx.GenerateRSAKeyPairWithLabel(id[:], label, teleport.RSAKeySize)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	key := keyID{
+		HostID: c.hostUUID,
+		KeyID:  id.String(),
+	}
+
+	keyID, err := key.marshal()
+	if err != nil {
+		return nil, nil, err
+	}
+	return keyID, signer, nil
+}
+
+// GetSigner returns a crypto.Signer for the given key identifier, if it is found.
+func (c *hsmKeyStore) GetSigner(rawKey []byte) (crypto.Signer, error) {
+	keyType := KeyType(rawKey)
+	switch keyType {
+	case types.PrivateKeyType_PKCS11:
+		keyID, err := parseKeyID(rawKey)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if keyID.HostID != c.hostUUID {
+			return nil, trace.NotFound("pkcs11 key is for different host")
+		}
+		pkcs11ID, err := keyID.pkcs11Key()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		signer, err := c.ctx.FindKeyPair(pkcs11ID, label)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if signer == nil {
+			return nil, trace.NotFound("failed to find keypair for given id")
+		}
+		return signer, nil
+	case types.PrivateKeyType_RAW:
+		return nil, trace.BadParameter("cannot get raw signer from HSM KeyStore")
+	}
+	return nil, trace.BadParameter("unrecognized key type %s", keyType.String())
+}
+
+func (c *hsmKeyStore) selectTLSKeyPair(ca types.CertAuthority) (*types.TLSKeyPair, error) {
```

Comment:
I don't really see a use for it outside of the `KeyStore`. I've tried to abstract away the "do I have an HSM or not" concept so that most code outside of this package does not need to care and we don't have conditionals everywhere. If it's needed elsewhere this can easily be refactored.